File: clients/kotlin/ridesharesdk/src/main/kotlin/rideshare/sdk/Driver.kt
package rideshare.sdk

import ares.client.fhe.CryptoContext
import kotlin.random.Random

object Driver {
    fun encryptBid(
        priceCents: Int, pkBytes: ByteArray,
        nonce: ByteArray = Random.nextBytes(16),
        ringDim: Int = 1 shl 15, depth: Int = 5,
        scalingFactor: Double = Math.scalb(1.0, 50)
    ): EncryptedBid {
        val ctx = CryptoContext(ringDim, scalingFactor, depth)
        ctx.use {
            val pk = ctx.deserializePublicKey(pkBytes)
            val ct = ctx.encrypt(doubleArrayOf(priceCents.toDouble(), 0.0, 0.0, 0.0), pk)
            return EncryptedBid(ctx.serialize(ct), priceCents, nonce)
        }
    }
}

File: clients/rideshareswift/Sources/RideshareCoreLib/Driver.swift
import AresClientFHE

public enum Driver {
    public static func encryptBid(priceCents: Int, under pkBytes: Data,
        ringDim: UInt32 = 1 << 15, depth: UInt32 = 5,
        scalingFactor: Double = Double(UInt64(1) << 50)) throws -> EncryptedBid {
        let ctx = try CryptoContext(ringDim: ringDim, scalingFactor: scalingFactor, depth: depth)
        let pk = try ctx.deserializePublicKey(Array(pkBytes))
        let ct = try ctx.encrypt([Double(priceCents), 0, 0, 0], under: pk)
        let ctBytes = try ctx.serialize(ct)
        let nonce = Data((0..<16).map { _ in UInt8.random(in: 0...255) })
        return EncryptedBid(ciphertext: Data(ctBytes), priceCents: priceCents, nonce: nonce)
    }
}
